core/src/main/java/org/apache/accumulo/core/iterators/SortedMapIterator.java

```diff
@@ -36,28 +37,40 @@
  *
  * Note that this class is intended as an in-memory replacement for RFile$Reader, so its behavior
  * reflects the same assumptions; namely, that this iterator is not responsible for respecting the
- * columnFamilies passed into seek().
+ * columnFamilies passed into seek(). If you want a Map-backed Iterator that returns only sought
+ * CFs, construct a new ColumnFamilySkippingIterator(new SortedMapIterator(map)).
+ *
+ * @see org.apache.accumulo.core.iterators.ColumnFamilySkippingIterator
  */
-public class SortedMapIterator implements SortedKeyValueIterator<Key,Value> {
+public class SortedMapIterator implements InterruptibleIterator {
   private Iterator<Entry<Key,Value>> iter;
   private Entry<Key,Value> entry;
 
   private SortedMap<Key,Value> map;
   private Range range;
 
+  private AtomicBoolean interruptFlag;
```

Comment:
Atomic references are typically `final` whenever possible, to prevent bugs where atomicity is broken because the reference is changed. I see this can't be made final because there's a setter... but that seems like bad API design. Is there something we can do to fix this?

Reply:
We should be able to make this final with an initial value of `false`. I haven't really touched the iterators since most of them have been around since forever but I think this would be a good fix. FYI I had a hell of a time tracking down where we actually do the interrupt. As a reference, it seems like the only place we actually interrupt iterators is in `ScanDataSource`:
https://github.com/apache/accumulo/blob/8a636a3ba91f5dae1d8b09b095178889a7d79c1d/server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/ScanDataSource.java#L256-L258
When a tablet is closed:
https://github.com/apache/accumulo/blob/b5ca15806659efbbd6f5d158be8648cc2cd13593/server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Tablet.java#L1288

We have a ton of calls to pass the flag around between iterators, readers and RFile code but that seems to be the only place where the actual interrupt takes place.

Reply:
I thought about changing the `AtomicBoolean` to be final but now I am wondering if it was written this way with performance in mind. We do a lot of passing around of the flag across the iterator stack but only ever set it in one place.  If we were to make it final, then I think that would be a lot of extra memory accesses to the atomic value vs just passing around a reference.

I thought of doing something like this in the setter, but like I said, I think it would be less efficient:
<pre>
public void setInterruptFlag(AtomicBoolean flag) {
    boolean newFlag = flag.get();
    boolean previousFlag = this.interruptFlag.getAndSet(newFlag);
    if (previousFlag == newFlag && newFlag)
      log.warn("Interrupt called on iterator that was previously interrupted.");
  }
</pre>

Reply:
It's not a performance thing. It's an implementation injection mechanism (see where we're notifying the iterators by changing the value in the scan code: `grep -lR 'interruptFlag.set(true)' .`. AtomicBoolean is being used like a BooleanSupplier. This was probably fine at one point, since BooleanSupplier was introduced in Java 8, and before that, AtomicBoolean was an attractive alternative to a built-in supplier. But, having the type be AtomicBoolean implies atomicity, and that can be confusing, since atomicity guarantees are clearly broken by having a setter that can swap out the reference itself in addition to the value. We're not actually relying on the atomicity guarantees in these cases, though, as far as I can tell, so BooleanSupplier would suffice as the type.

Your code above wouldn't work. I had a similar idea initially (`this.interruptFlag.set(flag.get())`). However, it's not sufficient to set the value. We need to set the containing object reference, so we can communicate the change in value when the caller updates the source.